src/v/redpanda/admin_server.cc

```diff
@@ -4632,3 +4644,73 @@ admin_server::restart_service_handler(std::unique_ptr<ss::http::request> req) {
     co_await restart_redpanda_service(*service);
     co_return ss::json::json_return_type(ss::json::json_void());
 }
+
+std::unique_ptr<ss::httpd::debug_json::memory_profile>
+get_sampled_memory_profile() {
+    auto stacks = ss::memory::sampled_memory_profile();
+    ss::sstring output;
+    for (const auto& stack : stacks) {
+        output += memory_sampling::format_allocation_site(stack);
+    }
+
+    // SS requires no-throw move constructable types but the json
+    // wrappers are not. Wrap in unique_ptr to hack around
+    auto profile = std::make_unique<ss::httpd::debug_json::memory_profile>();
+    profile->shard = ss::this_shard_id();
+    profile->profile = output;
+
+    return profile;
+}
+
+ss::future<ss::json::json_return_type>
+get_sampled_memory_profiles(std::optional<size_t> shard_id) {
+    using result_t = ss::httpd::debug_json::memory_profile;
+    std::vector<result_t> resp;
+
+    if (shard_id.has_value()) {
+        resp.push_back(
+          *co_await ss::smp::submit_to(*shard_id, get_sampled_memory_profile));
+    } else {
+        resp = co_await ss::map_reduce(
+          ss::smp::all_cpus(),
+          [](auto shard_id) -> ss::future<std::unique_ptr<result_t>> {
+              co_return co_await ss::smp::submit_to(
+                shard_id, get_sampled_memory_profile);
+          },
```

Comment:
this block of highlighted code looks like something that might be better to stick into the memory sampling service itself rather than in the admin web server.

separately:

```
        resp.push_back(
          *co_await ss::smp::submit_to(*shard_id, get_sampled_memory_profile));
    } else {
        resp = co_await ss::map_reduce(
          ss::smp::all_cpus(),
```

it's generally ideal to avoid raw `submit_to` calls when possible. this particular case looks like you might benefit from (1) sticking the memory sampling into a sharded<> service and (2) having that service inherit from `ss::peering_sharded_service` which will give it access to a `container()` local memory so you can implement things like `container().map_reduce` which is useful for reducing across all shards in the context of a service.

Reply:
Yep that sounds a lot cleaner.

I have moved this bit of code now to the memory_sampling service as suggested.

I haven't pulled in the API/JSON struct types directly into there as that feels a bit wrong. Means we have an intermittent carrier struct but that isn't even too bad as it means we can avoid the `unique_ptr` hack that was needed.

Reply:
yeh that looks pretty good

> Means we have an intermittent carrier struct 

yeh, this is the way to go. we aren't in a hot path or anything. and this is mostly strings so its probably copying a pointer and length anyway.